Took this for a spin and it appears to be working as described. Free trial:

<img width="1665" alt="Screenshot 2023-10-03 at 10 01 12 AM" src="https://github.com/Automattic/wp-calypso/assets/2124984/260b3b59-6962-419c-9483-afc81b4c92ed">

With the offer applied:

<img width="1664" alt="Screenshot 2023-10-03 at 10 02 07 AM" src="https://github.com/Automattic/wp-calypso/assets/2124984/f95a7ad7-904c-41bb-8422-10426bbd783b">

Had the offer applied and extended my plan:

<img width="1664" alt="Screenshot 2023-10-03 at 10 17 13 AM" src="https://github.com/Automattic/wp-calypso/assets/2124984/9b2c3a9c-b930-47e6-9834-44f173cf45d8">

With a plan but not the offer:

<img width="1662" alt="Screenshot 2023-10-03 at 10 28 58 AM" src="https://github.com/Automattic/wp-calypso/assets/2124984/f769b07f-df74-48e5-b32b-914213c38ed2">

Looks like there's a unit test failure: ```client/my-sites/plans-features-main/hooks/data-store/use-pricing-meta-for-grid-plans.ts(185,24): error TS2339: Property 'expiry' does not exist on type 'SitePlan'.```

--- a/packages/data-stores/src/plans/types.ts
+++ b/packages/data-stores/src/plans/types.ts
@@ -63,6 +63,8 @@ export interface SitePlan {
 	planSlug: PlanSlugFromProducts;
 	productId: number;
 	introOffer?: PlanIntroductoryOffer | null;
+	/* This value is only returned for the current plan on the site. */
+	expiry?: string;
 }
 
 export interface PricedAPIPlanIntroductoryOffer {

